packages/jest-core/src/SearchSource.ts

```diff
@@ -37,8 +37,19 @@ export type TestSelectionConfig = {
   watch?: boolean;
 };
 
-const globsToMatcher = (globs: Array<Config.Glob>) => (path: Config.Path) =>
-  micromatch([replacePathSepForGlob(path)], globs, {dot: true}).length > 0;
+let prevGlobs: Array<Config.Glob>;
+let prevGlobsMatchers: Array<(str: string) => boolean>;
+const globsToMatcher = (globs: Array<Config.Glob>) => {
+  if (prevGlobs !== globs) {
```

Comment:
have you tried caching that in a Map structure? does it impact memory too much?

Reply:
I didn't try because in my test runs `globs` was the same every time this function was called, so a Map would be unnecessary.

Reply:
@thymikee I think I've fixed this PR so tests don't fail now. The problem was that micromatch has some logic internally to handle lists of globs that may include negated globs and my naive approach of just checking if it matched any of the globs didn't capture that. I fixed this up by essentially copying and simplifying the logic from within micromatch. https://github.com/micromatch/micromatch/blob/fe4858b0c63b174fd3ae22674db39119b8fa4392/index.js#L27-L77

If this looks good, I think I'd like to land this, and then in https://github.com/facebook/jest/pull/10132 extract this code to a separate module that both places can use. I also think there are some other parts of the codebase that can make use of this for even more performance improvements (e.g. https://github.com/facebook/jest/blob/068ec04cc025596f01f79a0c472d9cc6553a72ca/packages/jest-transform/src/shouldInstrument.ts#L43), so having a reusable way to do this performantly seems like a good idea.